File: src/main/java/com/raychenon/hackerrank/warmup/SockMerchant.java
package com.raychenon.hackerrank.warmup;

import java.util.HashMap;
import java.util.Map;

/**
 * User: raychenon
 * Date: 8/4/19
 * https://www.hackerrank.com/challenges/sock-merchant/problem
 */
public class SockMerchant {

    int sockMerchant(int n, int[] ar) {
        Map<Integer, Integer> map = new HashMap<>();


        for (int i = 0; i < n; i++) {
            int value = ar[i];
            map.compute(value, (k, v) -> v == null ? 1 : v + 1);
        }
        System.out.println("Map = " + map);

        int pairSock = 0;
        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
            // only complete pairs are counted.
            // odd number are round down . ex: 3/2 = 1
            pairSock = pairSock + entry.getValue() / 2;
        }


        return pairSock;
    }

}

File: src/main/kotlin/com/raychenon/kotlin/hackerrank/warmup/SockMerchant.kt
package com.raychenon.kotlin.hackerrank.warmup

/**
 * User: raychenon
 * Date: 8/4/19
 * https://www.hackerrank.com/challenges/sock-merchant/problem
 */
object SockMerchant {

    fun sockMerchant(ar: Array<Int>): Int {
        val list = ar.toList()
        // count by frequency  ( key,frequency)
        val frequencies: Map<Int, Int> = list.groupingBy { it }.eachCount()

        // sum only the complete pairs in values
        val countPairSock = frequencies.values.sumBy { it -> it / 2 }

        return countPairSock

    }
}